crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -2820,6 +2745,121 @@ enum ForkedDependencies {
     },
 }
 
+/// A list of forks determined from the dependencies of a single package.
+///
+/// Any time a marker expression is seen that is not true for all possible
+/// marker environments, it is possible for it to introduce a new fork.
+#[derive(Debug, Default)]
+struct Forks {
+    /// The forks discovered among the dependencies.
+    forks: Vec<Fork>,
+    /// The package(s) that provoked at least one additional fork.
+    diverging_packages: Vec<PackageName>,
+}
+
+impl Forks {
+    /// Add the given dependency to this fork.
+    fn add(&mut self, dep: PubGrubDependency) {
+        use crate::marker::{is_definitively_empty_set, is_disjoint};
+
+        // A root can never be a dependency of another package,
+        // and a `Python` pubgrub package is never returned by
+        // `get_dependencies`. So a pubgrub package always has a
+        // name in this context.
+        let name = dep.package.name().expect("dependency always has a name");
+        let mut markers = dep
+            .package
+            .marker()
+            .cloned()
+            .unwrap_or(MarkerTree::And(vec![]));
+        if is_definitively_empty_set(&markers) {
+            // A marker that we know never matches any environment
+            // can be skipped entirely because its dependency will
+            // never be included in any circumstance.
+            return;
+        } else if self.forks.is_empty() {
+            // If we don't have any forks yet, we just seed it with
+            // two forks from `markers` and `not(markers)`.
+            if !markers.is_universal() {
+                self.diverging_packages.push(name.clone());
+                self.forks.push(Fork {
+                    dependencies: vec![],
+                    markers: markers.negate(),
+                });
+            }
+            self.forks.push(Fork {
+                dependencies: vec![dep],
+                markers,
+            });
+            return;
+        } else if markers.is_universal() {
+            // If our markers match every marker environment,
+            // then the dependency should be added to every fork.
+            for fork in &mut self.forks {
+                fork.dependencies.push(dep.clone());
+            }
+            return;
+        }
+        // Otherwise, we *may* need to add a new fork...
+
+        let mut new = vec![];
+        let mut found_overlap = false;
+        for mut fork in std::mem::take(&mut self.forks) {
```

Comment:
```suggestion
        for mut fork in std::mem::take(&mut self.forks) {
            // Invariant: The union of the markers of all forks is always the universe.
            // Let
            //   m_f := markers of the fork
            //   m_d := the remaining markers of the dependency
            // If `m_f ∩ m_d = ∅`, i.e. they are disjoint, we know that the dependency does not
            // occur in this fork, and we keep the fork as it was. Otherwise, `m_d` either
            // intersects with or covers `m_f`.
            // Should `m_f \ m_d` be empty, `m_d` covers `m_f`, and we add `m_f` verbatim and set
            // `m_d := m_d \ m_f`-
            // Otherwise, we split this fork into two new forks, `m_f \ m_d` and `m_f ∩ m_d`. This
            // is relevant if we have e.g. `python_version >= "3.9"` and `python_version >= "3.10"`,
            // so that we get forks `<3.9`, `>=3.9,<3.10` and `>=3.10`. It is a sound operation
            // since `(m_f \ m_d) ∪ (m_f ∩ m_d) = m_f` and since both new forks are either subsets
            // of `m_f` and `m_d`, meaning we miss no overlap.
```